tests/test_param_substitution.py

```diff
@@ -0,0 +1,215 @@
+  layer is to be removed.
+
+"""
+
+import os
+import shutil
+import copy
+import tempfile
+import unittest
+import glob
+from mock import patch
+from datetime import datetime
+from dateutil.relativedelta import relativedelta
+
+import in_toto.settings
+from in_toto.models.metadata import Metablock
+from in_toto.models.link import Link, FILENAME_FORMAT
+from in_toto.models.layout import (Step, Inspection, Layout,
+    SUBLAYOUT_LINK_DIR_FORMAT)
+from in_toto.verifylib import (verify_delete_rule, verify_create_rule,
+    verify_modify_rule, verify_allow_rule, verify_disallow_rule,
+    verify_match_rule, verify_item_rules, verify_all_item_rules,
+    verify_command_alignment, run_all_inspections, in_toto_verify,
+    verify_sublayouts, get_summary_link, _raise_on_bad_retval,
+    load_links_for_layout, verify_link_signature_thresholds,
+    verify_threshold_constraints, substitute_parameters)
+from in_toto.exceptions import (RuleVerificationError,
+    SignatureVerificationError, LayoutExpiredError, BadReturnValueError,
+    ThresholdVerificationError)
+from in_toto.util import import_rsa_key_from_file, import_public_keys_from_files_as_dict
+import in_toto.gpg.functions
+
+import securesystemslib.exceptions
+import in_toto.exceptions
+
+
+class Test_SubstituteArtifacts(unittest.TestCase):
+  """Test parameter substitution on artifact rules. """
+
+  def setUp(self):
+    self.layout = Layout.read({
+        "_type": "layout",
+        "inspect": [{
+          "name": "do-the-thing",
+          "expected_materials": [
+            ["MATCH", "{SOURCE_THING}", "WITH", "MATERIALS", "FROM", 
+              "{SOURCE_STEP}"]
+          ],
+          "expected_products": [
+            ["CREATE", "{NEW_THING}"]
+          ]
+        }],
+        "steps": [{
+          "name": "artifacts",
+          "expected_command": [],
+          "expected_materials": [
+            ["MATCH", "{SOURCE_THING}", "WITH", "MATERIALS", "FROM", 
+              "{SOURCE_STEP}"]
+          ],
+          "expected_products": [
+            ["CREATE", "{NEW_THING}"]
+          ]
+        }]
+    })
+
+  def test_substitute(self):
+    """Do a simple substitution on the expected_command field"""
+    substitute_parameters(self.layout, {"SOURCE_THING": "vim",
+      "SOURCE_STEP": "source_step", "NEW_THING": "new_thing"})
+    self.assertEquals(self.layout.steps[0].expected_materials[0][1], "vim")
+    self.assertEquals(self.layout.steps[0].expected_materials[0][5], "source_step")
+    self.assertEquals(self.layout.steps[0].expected_products[0][1], "new_thing")
+    self.assertEquals(self.layout.inspect[0].expected_materials[0][1], "vim")
+    self.assertEquals(self.layout.inspect[0].expected_materials[0][5], "source_step")
+    self.assertEquals(self.layout.inspect[0].expected_products[0][1], "new_thing")
+
+
+
+  def test_substitute_no_var(self):
+    """Raise an error if the parameter is not filled-in"""
+    with self.assertRaises(KeyError):
```

Comment:
Is there a unit test for a parameter dictionary that doesn't contain the expected format?

Another thing you  might want to do is raise a warning if the rule to be substituted doesn't contain the curly braces, which someone might forget to include.